cli/internal/run/real_run.go

```diff
@@ -89,30 +95,68 @@ func RealRun(
 		Concurrency: rs.Opts.runOpts.Concurrency,
 	}
 
-	mu := sync.Mutex{}
+	logMutex := sync.Mutex{}
+	logWaitGroup := sync.WaitGroup{}
+
+	outputLogsFunc := func(grouped bool, defaultOutWriter io.Writer, defaultErrWriter io.Writer, outBuf bytes.Buffer, errBuf bytes.Buffer) {
+		if grouped {
+			logMutex.Lock()
+			_, outErr := defaultOutWriter.Write(outBuf.Bytes())
+			_, errErr := defaultErrWriter.Write(errBuf.Bytes())
```

Comment:
This is going to result in out-of-order stdin and stdout message. I wonder if there's a way to maintain order between stdout and stderr?

Reply:
I thought about this as well. Unfortunately, I couldn't come up with a great solution.
The three solutions I thought of were:

1. Just sending both over stdout, and acting like stderr doesn't exist (stderr just gets forwarded to the outBuf). This maintains order, but you don't have any stderr being output anymore.

2. Custom implementation for io.Writer which forwards the content to a linked list, where we can preserve the order, as well as the origin. I can't make any promises about the performance of this though, maybe you have an idea of how performant this would be? A linked list would seem the most appropriate here, during writing we only care about the last element which we can keep a pointer to, during replaying we need to go through everything in order, which a linked list excels at.

3. Ignoring the order, but preserving the origin, as currently implemented.

Option 2 doesn't seem feasible, or worth the effort to me, however, if you believe it is feasible and worth the effort I would be happy to take a shot at implementing it. Option 1 would be about 5 minutes of work (including starting my docker dev environment) so if you prefer that over 3, just let me know.